### ISSUE TITLE
[Bug]: Yarn patches are not applied when node modules are loaded from a cache

### ISSUE THREAD

**Issue by milesj** [yarnpkg/berry#4231]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

Short: Yarn patches (`.yarn/patches`) are not applied when node modules are restored from a cache (ignores the fetch step).

Long: In CI, we cache the `.yarn/cache` and `.yarn/install-state.gz` files based on the checksum of `yarn.lock`. This works pretty great and reduces the "fetch step" during `yarn install` down to 1 second. Overall, the entirety of yarn install takes less than 2 minutes (most of it is linking).

However, we've noticed the new patches do not get applied when we use this approach, while patches work locally. This is very apparent because builds crash in CI that were fixed by this patch, and the line numbers in the stack trace are pointing to the original pre-patch lines.

Since we have _no way_ to apply patches manually, we have to bust our CI cache for this to work correctly.

### To reproduce

Our problem is unique to Buildkite CI but it can maybe be replicated with GitHub actions: https://github.com/actions/setup-node

### Environment

```shell
System:
    OS: macOS 12.3
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
  Binaries:
    Node: 16.13.1 - /private/var/folders/93/p1hnsld57s79d4b503vnmq2w0000gn/T/xfs-b98bb0af/node
    Yarn: 3.1.0 - /private/var/folders/93/p1hnsld57s79d4b503vnmq2w0000gn/T/xfs-b98bb0af/yarn
    npm: 8.1.2 - ~/.nvm/versions/node/v16.13.1/bin/npm
  npmPackages:
    jest: ^27.4.7 => 27.5.1
```

### Additional context

We're using Buildkite with the cache plugin:

```yaml
steps:
  - plugins:
    - <cache plugin>:
        key: yarn-node-modules-{{ checksum "yarn.lock" }}-v1
        paths:
          - ./.yarn/cache
          - ./.yarn/install-state.gz
```

**Comment by milesj** [comment]
I actually spoke too soon. Busting the cache doesn't fix the problem, and a fresh install doesn't seem to be patching either in CI. I'm looking into other issues.

**Comment by milesj** [comment]
I also tried removing the cache entirely and the files are still not patched in CI. This is _very_ confusing.

**Comment by milesj** [comment]
For example, this is the pre-patched code that fails:

```js
if (opts.platforms) {
    // @ts-ignore
    platforms = (0, _lodash().pick)(platforms, opts.platforms);

    _cliTools().logger.debug('Skipping selected platforms');
  }

  _cliTools().logger.debug('Available platforms: ' + `${Object.keys(platforms).map(_getPlatformName.default).join(', ')}`);

  if (rawPackageName === undefined) {
    _cliTools().logger.debug('No package name provided, will link all possible assets.');

    return (0, _linkAll.default)(ctx, {
      linkDeps: opts.all,
      linkAssets: true
    });
  }
```

The `return (0, _linkAll.default)(ctx, {` is line 82, and this is part of the error stack thrown in CI.

```
  at linkAll (/workdir/workspaces/node_modules/@react-native-community/cli/build/commands/link/linkAll.js:96:31)
  at Object.link [as func] (/workdir/workspaces/node_modules/@react-native-community/cli/build/commands/link/link.js:82:33)
```

And this is our patch:

```
diff --git a/build/commands/link/link.js b/build/commands/link/link.js
index 9f0468ba3b7b71885b8ed31415a763f54adc15f7..82085e0913b48464b0a77fb91150f6cb9b8c8f26 100644
--- a/build/commands/link/link.js
+++ b/build/commands/link/link.js
@@ -71,6 +71,10 @@ async function link([rawPackageName], ctx, opts) {
     // @ts-ignore
     platforms = (0, _lodash().pick)(platforms, opts.platforms);
 
+    // We filter down the platforms but dont apply it to the context,
+    // which results in ios still being targeted elsewhere.
+    ctx.platforms = platforms;
+
     _cliTools().logger.debug('Skipping selected platforms');
   }
```

And the patch applied to the code locally:

```js
if (opts.platforms) {
    // @ts-ignore
    platforms = (0, _lodash().pick)(platforms, opts.platforms);

    // We filter down the platforms but dont apply it to the context,
    // which results in ios still being targeted elsewhere.
    ctx.platforms = platforms;

    _cliTools().logger.debug('Skipping selected platforms');
  }

  _cliTools().logger.debug('Available platforms: ' + `${Object.keys(platforms).map(_getPlatformName.default).join(', ')}`);

  if (rawPackageName === undefined) {
    _cliTools().logger.debug('No package name provided, will link all possible assets.');

    return (0, _linkAll.default)(ctx, {
      linkDeps: opts.all,
      linkAssets: true
    });
  }
```

Line 82 is no longer `return (0, _linkAll.default)(ctx, {`, and is instead a black line 4 lines up.

Yet after all this, CI is still failing with the original error.

**Comment by milesj** [comment]
One piece of context that may be useful is that this is running in a docker container.

**Comment by milesj** [comment]
So in CI I went ahead and `cat` the file that should be patched after `yarn install`, and it definitely does not have the patch applied. Just going to switch to `patch-package`.

**Comment by milesj** [comment]
I also noticed that the lockfile isn't showing the patch defined in `resolutions`: `"@react-native-community/cli@6.4.0": "patch:@react-native-community/cli@npm:6.4.0#.yarn/patches/@react-native-community-cli-npm-6.4.0-f99ee73e09"`

**Comment by arcanis** [comment]
It's very difficult to understand what might happen without a reproduction, unfortunately. We use patches at work without issues, so there's something very specific to your context - if we can't isolate what that is, we can't help you much 🙁

**Comment by milesj** [comment]
@arcanis Yeah I figured. Creating a reproduction will be impossible.

Does Yarn have any kind of debugging I can enable or something?

**Comment by milesj** [comment]
Enabled `enableInlineHunks` to debug more and noticed TS is failing to apply. Would that stop other patches?

```
➤ YN0066: │ typescript@patch:typescript@npm%3A4.5.5#~builtin<compat/typescript>::version=4.5.5&hash=ddd1e8: Cannot apply hunk #2
--
  | ➤ YN0000: │           }
  | ➤ YN0000: │       }
  | ➤ YN0000: │       ts.getEffectiveTypeRoots = getEffectiveTypeRoots;
  | ➤ YN0028: │ -     function getDefaultTypeRoots(currentDirectory, host) {
  | ➤ YN0028: │ +     function getNodeModulesTypeRoots(currentDirectory, host) {
  | ➤ YN0000: │           if (!host.directoryExists) {
  | ➤ YN0000: │               return [ts.combinePaths(currentDirectory, nodeModulesAtTypes)];
  | ➤ YN0000: │           }
  | ➤ YN0028: │ -         var typeRoots;
  | ➤ YN0028: │ +         var typeRoots = [];
  | ➤ YN0000: │           ts.forEachAncestorDirectory(ts.normalizePath(currentDirectory), function (directory) {
  | ➤ YN0000: │               var atTypes = ts.combinePaths(directory, nodeModulesAtTypes);
  | ➤ YN0000: │               if (host.directoryExists(atTypes)) {
  | ➤ YN0028: │ -                 (typeRoots \|\| (typeRoots = [])).push(atTypes);
  | ➤ YN0028: │ +                 typeRoots.push(atTypes);
  | ➤ YN0000: │               }
  | ➤ YN0000: │               return undefined;
  | ➤ YN0000: │           });
  | ➤ YN0000: │           return typeRoots;
  | ➤ YN0000: │       }
  | ➤ YN0000: │       var nodeModulesAtTypes = ts.combinePaths("node_modules", "@types");
  | ➤ YN0028: │ +     function getPnpTypeRoots(currentDirectory) {
  | ➤ YN0028: │ +         var pnpapi = getPnpApi(currentDirectory);
  | ➤ YN0028: │ +         if (!pnpapi) {
  | ➤ YN0028: │ +             return [];
  | ➤ YN0028: │ +         }
  | ➤ YN0028: │ +         currentDirectory = ts.sys.resolvePath(currentDirectory);
  | ➤ YN0028: │ +         var currentPackage = pnpapi.findPackageLocator(currentDirectory + "/");
  | ➤ YN0028: │ +         if (!currentPackage) {
  | ➤ YN0028: │ +             return [];
  | ➤ YN0028: │ +         }
  | ➤ YN0028: │ +         var packageDependencies = pnpapi.getPackageInformation(currentPackage).packageDependencies;
  | ➤ YN0028: │ +         var typeRoots = [];
  | ➤ YN0028: │ +         for (var _i = 0, _a = Array.from(packageDependencies.entries()); _i < _a.length; _i++) {
  | ➤ YN0028: │ +             var _b = _a[_i], name = _b[0], referencish = _b[1];
  | ➤ YN0028: │ +             if (name.startsWith(typesPackagePrefix) && referencish !== null) {
  | ➤ YN0028: │ +                 var dependencyLocator = pnpapi.getLocator(name, referencish);
  | ➤ YN0028: │ +                 var packageLocation = pnpapi.getPackageInformation(dependencyLocator).packageLocation;
  | ➤ YN0028: │ +                 typeRoots.push(ts.getDirectoryPath(packageLocation));
  | ➤ YN0028: │ +             }
  | ➤ YN0028: │ +         }
  | ➤ YN0028: │ +         return typeRoots;
  | ➤ YN0028: │ +     }
  | ➤ YN0028: │ +     ts.getPnpTypeRoots = getPnpTypeRoots;
  | ➤ YN0028: │ +     var typesPackagePrefix = "@types/";
  | ➤ YN0028: │ +     function getDefaultTypeRoots(currentDirectory, host) {
  | ➤ YN0028: │ +         var nmTypes = getNodeModulesTypeRoots(currentDirectory, host);
  | ➤ YN0028: │ +         var pnpTypes = getPnpTypeRoots(currentDirectory);
  | ➤ YN0028: │ +         if (nmTypes.length > 0 \|\| pnpTypes.length > 0) {
  | ➤ YN0028: │ +             return __spreadArray(__spreadArray([], nmTypes, true), pnpTypes, true);
  | ➤ YN0028: │ +         }
  | ➤ YN0028: │ +     }
  | ➤ YN0000: │       function resolveTypeReferenceDirective(typeReferenceDirectiveName, containingFile, options, host, redirectedReference, cache) {
  | ➤ YN0000: │           var traceEnabled = isTraceEnabled(options, host);
  | ➤ YN0000: │           if (redirectedReference) {
```

We have a resolution entry for typescript, so we only have the 1 version.

**Comment by milesj** [comment]
@arcanis @merceyz Tried another repo, same problem. Had teammates test patches and they report it also not patching, even locally.

So I created a simple repo as a test case and the patch _still is not applied, even locally_: https://github.com/milesj/yarn-patch-transient-deps We tried Yarn 3.2 and 3.1 and both failed.

This looks to be a legitimate bug in Yarn.

**Comment by arcanis** [comment]
This particular problem is known (and fixed, once #4218 is merged). The `resolutions` syntax expects an exact match on the range compared to what the dependency truly is, but `yarn patch-commit` incorrectly sets a different range that generally doesn't match what the dependency actually is. For example in your example repo, `@react-native-community/cli` is depended upon by `react-native` which lists it with the range `"^6.0.0"` (not `"6.4.0"`).

To fixes that, just replace:

```json
"resolutions": {
  "@react-native-community/cli@6.4.0": "patch:..."
}
```

By:

```json
"resolutions": {
  "@react-native-community/cli@^6.0.0": "patch:..."
}
```

Another option is to remove the range altogether, in which case all instances of `@react-native-community/cli` will be patched (in your last message you said it doesn't work, but I suspect you made a mistaken copy-paste and forgot the range in your snippets, as I just tested it and it definitely works):

```json
"resolutions": {
  "@react-native-community/cli": "patch:..."
}
```

**Comment by milesj** [comment]
That... makes sense in retrospect. Will give that a try again on Monday.

Although the PR is improving patches, it would be nice if Yarn would warn if there's a patch that doesn't get applied for some reason.

**Comment by milesj** [comment]
Tested and works 👍 

Thanks for the info and reading my rambling.

**Comment by imranbarbhuiya** [comment]
> This particular problem is known (and fixed, once #4218 is merged). The `resolutions` syntax expects an exact match on the range compared to what the dependency truly is, but `yarn patch-commit` incorrectly sets a different range that generally doesn't match what the dependency actually is. For example in your example repo, `@react-native-community/cli` is depended upon by `react-native` which lists it with the range `"^6.0.0"` (not `"6.4.0"`).
> 
> To fixes that, just replace:
> 
> ```json
> "resolutions": {
>   "@react-native-community/cli@6.4.0": "patch:..."
> }
> ```
> 
> By:
> 
> ```json
> "resolutions": {
>   "@react-native-community/cli@^6.0.0": "patch:..."
> }
> ```
> 
> Another option is to remove the range altogether, in which case all instances of `@react-native-community/cli` will be patched (in your last message you said it doesn't work, but I suspect you made a mistaken copy-paste and forgot the range in your snippets, as I just tested it and it definitely works):
> 
> ```json
> "resolutions": {
>   "@react-native-community/cli": "patch:..."
> }
> ```

I'm still experiencing this issue and adding `^` manually works.

**Comment by sorenhoyer** [comment]
I also faced this issue today, can confirm that the `^` worked. It's quite a hard error to debug when using pnp, so I really hope it will get fixed soon, as I think many people will just give up before they come in here and realize how simple the fix is.

**Comment by xenoterracide** [comment]
@arcanis if mine is a duplicate of this, and this is closed, and mine happened today on the latest release, how can this be fixed/closed?

**Comment by arcanis** [comment]
Because it's in RCs for the next 4.x major, not in 3.x patches.

**Comment by imranbarbhuiya** [comment]
It's a bug fix so why it's not coming in 3x?

**Comment by xenoterracide** [comment]
Could this be Cherry picked back to 3.x please?

**Comment by wojtekmaj** [comment]
I'm still experiencing issues with my patch resolution on 4.0.0-rc.14.

These were the resolutions generated by patch-commit on 3.x:

```
"@rewardopl/react-native-toast@1.2.0": "patch:@rewardopl/react-native-toast@npm:1.2.0::__archiveUrl=https%3A%2F%2Fnpm.pkg.github.com%2Fdownload%2F%40rewardopl%2Freact-native-toast%2F1.2.0%2F30fa3fff8cce2eaaaff35fbc218041161631fd35cf06375f76834d890cffc2bd#.yarn/patches/@rewardopl-react-native-toast-npm-1.2.0-723ae2f371",
"@rewardopl/react-native-ui@1.7.0": "patch:@rewardopl/react-native-ui@npm:1.7.0::__archiveUrl=https%3A%2F%2Fnpm.pkg.github.com%2Fdownload%2F%40rewardopl%2Freact-native-ui%2F1.7.0%2Fe3c4f7ab023dac67d7ad2157fb6502a8cf46ac5f7f7b035f59f2c28c7a824264#.yarn/patches/@rewardopl-react-native-ui-npm-1.7.0-7602a18b2b",
```

* Changing key to `@rewardopl/react-native-toast` continues to throw an error. Note: unscoped packages didn't have this problem even in 3.x.
* Changing key to `@rewardopl/react-native-toast@^1.2.0` does not crash during install, but still doesn't apply the patch. No error or warning is visible during install.

**Comment by milesj** [comment]
Yes this is is still _very_ annoying. Anytime we use patches, they don't work immediately, and we constantly have to tinker with the resolution version to get them working. A fix in a release would be much appreciated.

**Comment by ffxsam** [comment]
When will this be released on the stable branch?

**Comment by merceyz** [comment]
> When will this fix be released on the stable branch?

The fix has been released in v3.2.3

**Comment by adam-thomas-privitar** [comment]
Can confirm this is still broken as above

**Comment by danguilherme** [comment]
> > When will this fix be released on the stable branch?
> 
> The fix has been released in v3.2.3

I am using yarn v3.3.1 and it didn't work for me.

**Comment by ps-leocxy** [comment]
> > > When will this fix be released on the stable branch?
> > 
> > 
> > The fix has been released in v3.2.3
> 
> I am using yarn v3.3.1 and it didn't work for me.

Same here...

**Comment by jleach** [comment]
I made the above change **and** found similar resolution higher up that was problematic in my case. Although for me the problem was it not applying all the patches, just some of them. Anyway, hope this helps...

```json
"resolutions": {
    "@types/react": "17.0.2",
    "@types/react-dom": "17.0.2",
    "@aries-framework/core": "0.3.3",
  ...
    "@aries-framework/core@0.3.3": "patch:@aries-framework/core@npm:0.3.3#./.yarn/patches/@aries-framework-core-npm-0.3.3-dd6486de3d.patch",
  }
```

I had to remove `"@aries-framework/core": "0.3.3",` for mine to work.

**Comment by odinho** [comment]
Something is definitely easy to get wrong here, `yarn` is *very* eager to create a *ton* of resolutions for every sub-version, instead of realizing it should just change the existing resolutions setting.

I've added the `resolution` on top specifically since I wanted /one/ patch to patch all minor versions in babel. Since the code is the same anyway. And that is likely where everyone is stumbling.

This is what yarn generated (after I manually added the version resolutions to only have to patch this once, as it hints to me when I use `yarn patch @babel/helpers`):
```
  "resolutions": {
    "@babel/helpers": "7.23.5",
    "@babel/runtime": "7.23.4",
    "@babel/helpers@^7.22.15": "patch:@babel/helpers@npm%3A7.23.5#./.yarn/patches/@babel-helpers-npm-7.23.5-3d1b3cb6c2.patch",
    "@babel/helpers@^7.23.5": "patch:@babel/helpers@npm%3A7.23.5#./.yarn/patches/@babel-helpers-npm-7.23.5-3d1b3cb6c2.patch",
    "@babel/helpers@^7.23.2": "patch:@babel/helpers@npm%3A7.23.5#./.yarn/patches/@babel-helpers-npm-7.23.5-3d1b3cb6c2.patch",
    "@babel/helpers@^7.22.6": "patch:@babel/helpers@npm%3A7.23.5#./.yarn/patches/@babel-helpers-npm-7.23.5-3d1b3cb6c2.patch",
    "@babel/helpers@^7.22.5": "patch:@babel/helpers@npm%3A7.23.5#./.yarn/patches/@babel-helpers-npm-7.23.5-3d1b3cb6c2.patch",
    "@babel/runtime@^7.21.0": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.14.6": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.4.2": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@7.12.1": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.20.7": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.5.5": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.12.5": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.8.7": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.23.4": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.13.10": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.12.0": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.10.2": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.9.2": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.15.4": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.17.8": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch",
    "@babel/runtime@^7.7.2": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch"
  },
```

This is what I changed it to:
```
  "resolutions": {
    "@babel/helpers": "patch:@babel/helpers@npm%3A7.23.5#./.yarn/patches/@babel-helpers-npm-7.23.5-3d1b3cb6c2.patch",
    "@babel/runtime": "patch:@babel/runtime@npm%3A7.23.4#./.yarn/patches/@babel-runtime-npm-7.23.4-2e68957572.patch"
  },
```

So `yarn` should probably do that by itself. When the package it is patching has a resolutions field, it should overwrite that with the `patch:` it has created.

**Comment by movy** [comment]
just encountered this with the most recent yarn 4.0.2, after fiddling with package resolutions, the patch was applied on production, but still could not see it on staging CI, very weird. No error messages either.

Initial package.json:

```json
    "@uppy/compressor": "patch:@uppy/compressor@npm%3A1.0.5#~/.yarn/patches/@uppy-compressor-npm-1.0.5-c6622da068.patch",

```
What worked:
```json
    "@uppy/compressor": "patch:@uppy/compressor@npm:^1.0.5#../../../../.yarn/patches/@uppy-compressor-npm-1.0.5-c6622da068.patch",
```

(unescaped `%3A` to `:`, added `^` and changed path to relative)

yarn.lock:
```json
"@uppy/compressor@npm:^1.0.5":
  version: 1.0.5
  resolution: "@uppy/compressor@npm:1.0.5"
  dependencies:
    "@transloadit/prettier-bytes": "npm:^0.0.9"
    "@uppy/utils": "npm:^5.5.2"
    compressorjs: "npm:^1.1.1"
    preact: "npm:^10.5.13"
    promise-queue: "npm:^2.2.5"
  peerDependencies:
    "@uppy/core": ^3.6.0
  checksum: bd550351845b5d75daaf7f9626dce2f49a4c9a51dd9e665e05b2748ae8fd4d4f83c5973ec7aed43df9eea5053264f9bec3e167a622228f038b7a6dd4d341e298
  languageName: node
  linkType: hard

"@uppy/compressor@patch:@uppy/compressor@npm:^1.0.5#../../../../.yarn/patches/@uppy-compressor-npm-1.0.5-c6622da068.patch::locator=vue%40workspace%3Afront%2Fweb_server%2Fwww%2F_vue":
  version: 1.0.5
  resolution: "@uppy/compressor@patch:@uppy/compressor@npm%3A1.0.5#../../../../.yarn/patches/@uppy-compressor-npm-1.0.5-c6622da068.patch::version=1.0.5&hash=58663d&locator=vue%40workspace%3Afront%2Fweb_server%2Fwww%2F_vue"
  dependencies:
    "@transloadit/prettier-bytes": "npm:^0.0.9"
    "@uppy/utils": "npm:^5.5.2"
    compressorjs: "npm:^1.1.1"
    preact: "npm:^10.5.13"
    promise-queue: "npm:^2.2.5"
  peerDependencies:
    "@uppy/core": ^3.6.0
  checksum: a79efe83d7bf503551f5d293569d094438d7b62cb7941624687f9100a240936debf5a84016c8348f2e8e8f1db66da8d8b17628a22ee349910bdfae5b9fef3075
  languageName: node
  linkType: hard

```

**Comment by gozegotushar** [comment]
i was also facing problem with yarn : 3.6.4 
yarn install was not successful, for other team mates yarn install was working fine.
--- 
![Screenshot 2024-09-06 at 6 58 12 PM (1)](https://github.com/user-attachments/assets/5cd40a04-1a5a-4ba3-bbe3-4f18541204eb)
---
solution used : 

- removed ~/ before yarn path.
c48c867a2c91)

**getting issue with** :     `"react-native-root-toast": "patch:react-native-root-toast@npm%3A3.5.1#~/.yarn/patches/react-native-root-toast-npm-3.5.1-f04b46d674.patch"`,
**issue fixed** :    `"react-native-root-toast": "patch:react-native-root-toast@npm%3A3.5.1#.yarn/patches/react-native-root-toast-npm-3.5.1-f04b46d674.patch"`,

- used locator with upgraded patch.

**getting issue with** :     `"react-native-snap-carousel": "patch:react-native-snap-carousel@patch%3Areact-native-snap-carousel@npm%253A3.9.1%23~/.yarn/patches/react-native-snap-carousel-npm-3.9.1-14bebfe4a7.patch%3A%3Aversion=3.9.1&hash=25be1a#~/.yarn/patches/react-native-snap-carousel-patch-c304848e43.patch"`,    
**issue fixed** : `"react-native-snap-carousel": "patch:react-native-snap-carousel@patch%3Areact-native-snap-carousel@npm%253A3.9.1%23.yarn/patches/react-native-snap-carousel-npm-3.9.1-14bebfe4a7.patch%3A%3Aversion=3.9.1&hash=25be1a&locator=resident_app%2540workspace%253A.#.yarn/patches/react-native-snap-carousel-patch-c304848e43.patch"`
